doc/decisions/0001-define-piping-in-a-structured-way.md

```diff
@@ -0,0 +1,310 @@
+                        "placeholders": [
+                            {
+                                "placeholder": "person_name",
+                                "transforms": [
+                                    {
+                                        "transform": "concatenate_list",
+                                        "arguments": {
+                                            "list": {
+                                                "source": "answers",
+                                                "identifier": ["first-name","last-name"]
+                                            },
+                                            "delimiter": " "
+                                        }
+                                    }
+                                ]
+                            },
+                            {
+                                "placeholder": "age_in_years",
+                                "transforms": [
+                                    {
+                                        "transform": "calculate_years_difference",
+                                        "arguments": {
+                                            "first_date": {
+                                                "source": "answers",
+                                                "identifier": "date-of-birth-answer"
+                                            },
+                                            "second_date": {
+                                                "value": "now"
+                                            }
+                                        }
+                                    }
+                                ]
+                            }
+                        ]
+                    },
+                    "value": "Yes"
+                },
+                {
+                    "label": {
+                        "text": "No, I need to change their date of birth"
+                    },
+                    "value": "No"
+                }
+            ],
+            "type": "Radio"
+        }]
+    }]
+}
+```
+
+Note: The repetition of placeholder resolution in the same block is rare
+
+---
+
+
+## Other notes
+
+- No provision has been made in the schema design for group instances. It is assumed that:
+  - When a placeholder is used in a non-repeating group that any reference to answers resolves to all answers that match the answer id.
+  - When a placeholder is used in a repeating group that any reference to answers resolves to the answer that matches within the current repeat.
+- `format_date_range` and `format_datetime` should be done via two separate placeholders rather than a transform, so that the 'to' and 'at' are in the sentence to be translated.
+- Jinja filters are used to resolve values in the variables schema e.g. `period` in 1_0005. This will no longer work and will require some further design outside of the scope of this proposal.
```

Comment:
I think all of these examples look good. The structured piping approach you've taken is similar to how we've discussed modelling piping in author so we may revisit our implementation at some point and borrow some ideas from here. 👍

There's one scenario that we have a requirement for that I don't see an example of here. Wondering if this would just be another type of transform with two args?

`{{ one_of(metadata['trad_as'], metadata['ru_name']) }}`

Reply:
That's good to hear 🙂

For the `one_of` scenario there is a `first_non_empty_item` filter in runner that takes a list of values and returns the first non-empty one. In new structured json it could look something like this:

```json
"transforms": [
    {
        "transform": "first_non_empty_item",
        "arguments": {
            "items": {
                "source": "metadata",
                "identifier": ["ref_p_start_date","ru_name"]
            },
        }
    }
]
```

or (this would allow for each item to come from a different source):

```json
"transforms": [
    {
        "transform": "first_non_empty_item",
        "arguments": {
            "items": [
				{
                	"source": "metadata",
	                "identifier": "ref_p_start_date"
				},
				{
                	"source": "metadata",
	                "identifier": "ru_name"
				}
            ]
        }
    }
]
```

Reply:
Nice, yeah that makes sense